File: 198.打家劫舍.cpp
/*
 * @lc app=leetcode.cn id=198 lang=cpp
 *
 * [198] 打家劫舍
 */

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums.front();
        auto [dp0, dp1] = pair<int, int>({nums[0], max(nums[0], nums[1])});
        for(int i=2; i<nums.size(); ++i){
            tie(dp0, dp1) = pair<int, int>({dp1, max(dp1, dp0 + nums[i])});
        }
        return dp1;
    }
};
// @lc code=end

File: 198.打家劫舍.py
#
# @lc app=leetcode.cn id=198 lang=python3
#
# [198] 打家劫舍
#

# @lc code=start
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1: return nums[0]
        dp = [0] * len(nums)
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-2] + nums[i], dp[i-1])
        return dp[-1]
# @lc code=end

print(Solution().rob([1,2,3,1]))
print(Solution().rob([2,7,9,3,1]))